packages/dd-trace/test/plugins/promise.js

```diff
@@ -12,95 +13,129 @@ module.exports = (name, factory) => {
+              const promise = new Promise((resolve, reject) => {
+                setImmediate(() => {
+                  tracer.scope().activate({}, () => {
+                    reject(new Error())
+                  })
                 })
               })
+
+              return tracer.scope().activate(span, () => {
+                return promise
+                  .catch(err => {
+                    throw err
+                  })
+                  .catch(() => {
+                    expect(tracer.scope().active()).to.equal(span)
+                  })
+              })
             })
 
-            return tracer.scope().activate(span, () => {
-              return promise
-                .catch(err => {
-                  throw err
-                })
-                .catch(() => {
-                  expect(tracer.scope().active()).to.equal(span)
+            it('should allow to run without a scope if not available when calling then()', () => {
+              if (process.env.DD_CONTEXT_PROPAGATION === 'false') return
+
+              tracer.scope().activate(null, () => {
+                const promise = new Promise((resolve, reject) => {
+                  setImmediate(() => {
+                    tracer.scope().activate({}, () => {
+                      resolve()
+                    })
+                  })
                 })
+
+                return promise
+                  .then(() => {
+                    expect(tracer.scope().active()).to.be.null
+                  })
+              })
             })
-          })
 
-          it('should allow to run without a scope if not available when calling then()', () => {
-            if (process.env.DD_CONTEXT_PROPAGATION === 'false') return
+            it('should unpatch then() callback when unpatching instrumentation', () => {
+              if (process.env.DD_CONTEXT_PROPAGATION === 'false') return
 
-            tracer.scope().activate(null, () => {
-              const promise = new Promise((resolve, reject) => {
+              tracer._instrumenter.disable()
+
+              const span = {}
+              const differentContextSpan = {}
+
+              let promise = new Promise((resolve, reject) => {
                 setImmediate(() => {
-                  tracer.scope().activate({}, () => {
+                  tracer.scope().activate(differentContextSpan, () => {
                     resolve()
                   })
                 })
               })
 
-              return promise
-                .then(() => {
-                  expect(tracer.scope().active()).to.be.null
-                })
+              return tracer.scope().activate(span, () => {
+                for (let i = 0; i < promise.then.length; i++) {
+                  const args = new Array(i + 1)
+
+                  args[i] = () => {
+                    expect(tracer.scope().active()).to.equal(differentContextSpan)
+                    expect(span).to.not.equal(differentContextSpan)
```

Comment:
This could fail if the library uses `async_hooks` internally. A safer check would be to check that the behavior is the same before patching and after unpatching.

Reply:
thanks @rochdev , that makes sense at a high level.  I've updated this to test, what I believe, to be the best way i can measure before / after patching. I had to do a bit of re-structuring of the test suite and it gave me a bit of trouble to structure the test in a sane way and it's still a little bit hacky but I think this ought to be a scalable approach for broader promise libraries. lmk what you think